pkg/controller/master/node/node_volume_detach_controller.go

```diff
@@ -0,0 +1,190 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	ctlcorev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/harvester/harvester/pkg/config"
+	ctllonghornv1 "github.com/harvester/harvester/pkg/generated/controllers/longhorn.io/v1beta1"
+	ctlsnapshotv1 "github.com/harvester/harvester/pkg/generated/controllers/snapshot.storage.k8s.io/v1beta1"
+	"github.com/harvester/harvester/pkg/indexeres"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	nodeVolumeDetachControllerName  = "node-volume-detach-controller"
+	nodeVolumeDetachEnqueueInterval = 5 * time.Second
+)
+
+// nodeVolumeDetachController is responsible for detaching volumes from a node when the node is in draining
+// fix https://github.com/harvester/harvester/issues/3681
+type nodeVolumeDetachController struct {
+	nodeController ctlcorev1.NodeController
+	podCache       ctlcorev1.PodCache
+	pvcCache       ctlcorev1.PersistentVolumeClaimCache
+	volumes        ctllonghornv1.VolumeClient
+	volumeCache    ctllonghornv1.VolumeCache
+	snapshotCache  ctlsnapshotv1.VolumeSnapshotCache
+}
+
+// VolumeDetachRegister registers the node volume detach controller
+func VolumeDetachRegister(ctx context.Context, management *config.Management, options config.Options) error {
+	nodes := management.CoreFactory.Core().V1().Node()
+	pods := management.CoreFactory.Core().V1().Pod()
+	pvcs := management.CoreFactory.Core().V1().PersistentVolumeClaim()
+	volumes := management.LonghornFactory.Longhorn().V1beta1().Volume()
+	snapshots := management.SnapshotFactory.Snapshot().V1beta1().VolumeSnapshot()
+	nodeVolumeDetachController := &nodeVolumeDetachController{
+		nodeController: nodes,
+		podCache:       pods.Cache(),
+		pvcCache:       pvcs.Cache(),
+		volumes:        volumes,
+		volumeCache:    volumes.Cache(),
+		snapshotCache:  snapshots.Cache(),
+	}
+
+	nodes.OnChange(ctx, nodeVolumeDetachControllerName, nodeVolumeDetachController.OnNodeChanged)
+
+	return nil
+}
+
+// OnNodeChanged is called when a node is changed
+func (c *nodeVolumeDetachController) OnNodeChanged(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+
+	// skip if node is not in draining
```

Comment:
add further check of upgrade context to make sure this is only working in upgrade ?